File: database/database-annotation-processor/src/main/java/ru/tinkoff/kora/database/annotation/processor/model/QueryResult.java
package ru.tinkoff.kora.database.annotation.processor.model;

import com.squareup.javapoet.TypeName;
import ru.tinkoff.kora.annotation.processor.common.CommonUtils;
import ru.tinkoff.kora.annotation.processor.common.ProcessingErrorException;

public sealed interface QueryResult {
    TypeName type();

    interface ReactiveResult {}

    record SimpleResult(TypeName type) implements QueryResult {}

    record ResultWithMapper(TypeName type, CommonUtils.MappingData mappingData) implements QueryResult {}

    record MonoResult(TypeName type, QueryResult result) implements QueryResult, ReactiveResult {
        public MonoResult {
            if (!(result instanceof SimpleResult || result instanceof ResultWithMapper)) {
                throw new ProcessingErrorException("Invalid mono type", null);
            }
        }
    }
    record FluxResult(TypeName type, QueryResult result) implements QueryResult, ReactiveResult {
        public FluxResult {
            if (!(result instanceof SimpleResult || result instanceof ResultWithMapper)) {
                throw new ProcessingErrorException("Invalid flux type", null);
            }
        }
    }
}

File: database/database-symbol-processor/src/main/kotlin/ru/tinkoff/kora/database/symbol/processor/model/QueryResult.kt
package ru.tinkoff.kora.database.symbol.processor.model

import com.google.devtools.ksp.symbol.KSType
import ru.tinkoff.kora.ksp.common.MappingData
import ru.tinkoff.kora.ksp.common.exception.ProcessingErrorException

sealed interface QueryResult {
    val type: KSType

    interface ReactiveResult

    data class SimpleResult(override val type: KSType) : QueryResult

    data class ResultWithMapper constructor(override val type: KSType, val mappingData: MappingData) : QueryResult

    data class SuspendResult(override val type: KSType, val result: QueryResult) : QueryResult, ReactiveResult {
        init {
            if (!(result is SimpleResult || result is ResultWithMapper)) {
                throw ProcessingErrorException("Invalid suspend type: $result", type.declaration)
            }
        }
    }
}
